File: go/annotations/parser.go
package annotations

import (
	"fmt"
	"go/ast"
	"go/parser"
	"go/token"
	"os"
	"path/filepath"
	"strings"
	"time"
)

// Parser handles parsing of Go source files for annotations
type Parser struct {
	fset *token.FileSet
}

// NewParser creates a new annotation parser
func NewParser() *Parser {
	return &Parser{
		fset: token.NewFileSet(),
	}
}

// ParseDirectory parses all Go files in a directory recursively
func (p *Parser) ParseDirectory(dir string) (*ParsedAnnotations, error) {
	result := &ParsedAnnotations{
		Handlers: make([]Handler, 0),
		Errors:   make([]ParseError, 0),
	}

	// Walk the directory tree
	err := filepath.Walk(dir, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return err
		}

		// Skip directories and non-Go files
		if info.IsDir() || !strings.HasSuffix(path, ".go") {
			return nil
		}

		// Skip test files
		if strings.HasSuffix(path, "_test.go") {
			return nil
		}

		// Parse the file
		fileResult, err := p.ParseFile(path)
		if err != nil {
			result.Errors = append(result.Errors, ParseError{
				FilePath: path,
				Message:  fmt.Sprintf("Failed to parse file: %v", err),
			})
			return nil // Continue parsing other files
		}

		// Merge results
		result.Handlers = append(result.Handlers, fileResult.Handlers...)
		result.Errors = append(result.Errors, fileResult.Errors...)

		return nil
	})

	if err != nil {
		return nil, fmt.Errorf("failed to walk directory: %w", err)
	}

	return result, nil
}

// ParseFile parses a single Go file for annotations
func (p *Parser) ParseFile(filePath string) (*ParsedAnnotations, error) {
	result := &ParsedAnnotations{
		Handlers: make([]Handler, 0),
		Errors:   make([]ParseError, 0),
	}

	// Parse the file
	file, err := parser.ParseFile(p.fset, filePath, nil, parser.ParseComments)
	if err != nil {
		return nil, fmt.Errorf("failed to parse file: %w", err)
	}

	// Get absolute path
	absPath, err := filepath.Abs(filePath)
	if err != nil {
		absPath = filePath
	}

	// Extract package name
	packageName := file.Name.Name

	// Find all function declarations with annotations
	ast.Inspect(file, func(n ast.Node) bool {
		// Look for function declarations
		funcDecl, ok := n.(*ast.FuncDecl)
		if !ok {
			return true
		}

		// Skip methods (we only want top-level functions)
		if funcDecl.Recv != nil {
			return true
		}

		// Get function name
		funcName := funcDecl.Name.Name

		// Get function position
		position := p.fset.Position(funcDecl.Pos())

		// Extract annotations from doc comments
		if funcDecl.Doc == nil {
			return true // No comments, skip
		}

		// Parse annotations from comments
		handler, parseErrs := p.parseAnnotations(funcDecl.Doc, funcName, packageName, absPath, position.Line)

		if handler != nil {
			result.Handlers = append(result.Handlers, *handler)
		}

		result.Errors = append(result.Errors, parseErrs...)

		return true
	})

	return result, nil
}

// parseAnnotations extracts @wylla:* annotations from comment group
func (p *Parser) parseAnnotations(doc *ast.CommentGroup, funcName, packageName, filePath string, lineNumber int) (*Handler, []ParseError) {
	handler := &Handler{
		FunctionName: funcName,
		PackageName:  packageName,
		FilePath:     filePath,
		LineNumber:   lineNumber,
		Auth: AuthConfig{
			Type: AuthNone, // Default to no auth
		},
	}

	var errors []ParseError
	hasBoxAnnotation := false

	// Process each comment line
	for _, comment := range doc.List {
		text := strings.TrimSpace(comment.Text)

		// Remove comment markers
		text = strings.TrimPrefix(text, "//")
		text = strings.TrimPrefix(text, "/*")
		text = strings.TrimSuffix(text, "*/")
		text = strings.TrimSpace(text)

		// Check if it's a Box annotation
		if !strings.HasPrefix(text, "@box:") {
			continue
		}

		hasBoxAnnotation = true

		// Parse the annotation
		parts := strings.SplitN(text, ":", 2)
		if len(parts) != 2 {
			errors = append(errors, ParseError{
				FilePath:   filePath,
				LineNumber: lineNumber,
				Message:    fmt.Sprintf("Invalid annotation format: %s", text),
				Annotation: text,
			})
			continue
		}

		annotationType := strings.TrimSpace(parts[1])
		var annotationValue string

		// Split annotation type and value
		if spaceIdx := strings.Index(annotationType, " "); spaceIdx > 0 {
			annotationValue = strings.TrimSpace(annotationType[spaceIdx+1:])
			annotationType = annotationType[:spaceIdx]
		}

		// Parse based on annotation type
		switch annotationType {
		case "function":
			handler.DeploymentType = DeploymentFunction

		case "container":
			handler.DeploymentType = DeploymentContainer
			// Parse optional service=name parameter
			if annotationValue != "" {
				if err := p.parseContainerService(handler, annotationValue); err != nil {
					errors = append(errors, ParseError{
						FilePath:   filePath,
						LineNumber: lineNumber,
						Message:    fmt.Sprintf("Invalid container annotation: %v", err),
						Annotation: text,
					})
				}
			}

		case "path":
			if err := p.parsePath(handler, annotationValue); err != nil {
				errors = append(errors, ParseError{
					FilePath:   filePath,
					LineNumber: lineNumber,
					Message:    fmt.Sprintf("Invalid path annotation: %v", err),
					Annotation: text,
				})
			}

		case "auth":
			if err := p.parseAuth(handler, annotationValue); err != nil {
				errors = append(errors, ParseError{
					FilePath:   filePath,
					LineNumber: lineNumber,
					Message:    fmt.Sprintf("Invalid auth annotation: %v", err),
					Annotation: text,
				})
			}

		case "ratelimit":
			if err := p.parseRateLimit(handler, annotationValue); err != nil {
				errors = append(errors, ParseError{
					FilePath:   filePath,
					LineNumber: lineNumber,
					Message:    fmt.Sprintf("Invalid ratelimit annotation: %v", err),
					Annotation: text,
				})
			}

		case "cors":
			if err := p.parseCORS(handler, annotationValue); err != nil {
				errors = append(errors, ParseError{
					FilePath:   filePath,
					LineNumber: lineNumber,
					Message:    fmt.Sprintf("Invalid cors annotation: %v", err),
					Annotation: text,
				})
			}

		case "timeout":
			if err := p.parseTimeout(handler, annotationValue); err != nil {
				errors = append(errors, ParseError{
					FilePath:   filePath,
					LineNumber: lineNumber,
					Message:    fmt.Sprintf("Invalid timeout annotation: %v", err),
					Annotation: text,
				})
			}

		case "memory":
			handler.Memory = annotationValue

		case "concurrency":
			var concurrency int
			if _, err := fmt.Sscanf(annotationValue, "%d", &concurrency); err != nil {
				errors = append(errors, ParseError{
					FilePath:   filePath,
					LineNumber: lineNumber,
					Message:    fmt.Sprintf("Invalid concurrency value: %s", annotationValue),
					Annotation: text,
				})
			} else {
				handler.Concurrency = concurrency
			}

		default:
			errors = append(errors, ParseError{
				FilePath:   filePath,
				LineNumber: lineNumber,
				Message:    fmt.Sprintf("Unknown annotation type: %s", annotationType),
				Annotation: text,
			})
		}
	}

	// If no Box annotations found, return nil
	if !hasBoxAnnotation {
		return nil, errors
	}

	return handler, errors
}

// parsePath parses @box:path METHOD /path/to/resource
func (p *Parser) parsePath(handler *Handler, value string) error {
	parts := strings.SplitN(value, " ", 2)
	if len(parts) != 2 {
		return fmt.Errorf("path must be in format 'METHOD /path', got: %s", value)
	}

	method := strings.ToUpper(strings.TrimSpace(parts[0]))
	path := strings.TrimSpace(parts[1])

	// Validate HTTP method
	validMethods := map[string]bool{
		"GET": true, "POST": true, "PUT": true, "DELETE": true,
		"PATCH": true, "OPTIONS": true, "HEAD": true,
	}
	if !validMethods[method] {
		return fmt.Errorf("invalid HTTP method: %s", method)
	}

	// Validate path starts with /
	if !strings.HasPrefix(path, "/") {
		return fmt.Errorf("path must start with /, got: %s", path)
	}

	handler.Route = Route{
		Method: method,
		Path:   path,
	}

	return nil
}

// parseAuth parses @wylla:auth required|optional|none
func (p *Parser) parseAuth(handler *Handler, value string) error {
	value = strings.ToLower(strings.TrimSpace(value))

	switch value {
	case "required":
		handler.Auth.Type = AuthRequired
	case "optional":
		handler.Auth.Type = AuthOptional
	case "none":
		handler.Auth.Type = AuthNone
	default:
		return fmt.Errorf("auth must be 'required', 'optional', or 'none', got: %s", value)
	}

	return nil
}

// parseRateLimit parses @wylla:ratelimit 100/hour
func (p *Parser) parseRateLimit(handler *Handler, value string) error {
	parts := strings.Split(value, "/")
	if len(parts) != 2 {
		return fmt.Errorf("ratelimit must be in format 'count/period', got: %s", value)
	}

	var count int
	if _, err := fmt.Sscanf(parts[0], "%d", &count); err != nil {
		return fmt.Errorf("invalid count in ratelimit: %s", parts[0])
	}

	period := strings.ToLower(strings.TrimSpace(parts[1]))
	var duration int64

	switch period {
	case "second", "sec", "s":
		duration = 1
	case "minute", "min", "m":
		duration = 60
	case "hour", "hr", "h":
		duration = 3600
	case "day", "d":
		duration = 86400
	default:
		return fmt.Errorf("invalid period in ratelimit: %s (use second/minute/hour/day)", period)
	}

	handler.RateLimit = &RateLimitConfig{
		Count:  count,
		Period: parseDuration(duration),
		Raw:    value,
	}

	return nil
}

// parseCORS parses @wylla:cors origins=*
func (p *Parser) parseCORS(handler *Handler, value string) error {
	if !strings.HasPrefix(value, "origins=") {
		return fmt.Errorf("cors must be in format 'origins=*' or 'origins=url1,url2', got: %s", value)
	}

	originsStr := strings.TrimPrefix(value, "origins=")
	var origins []string

	if originsStr == "*" {
		origins = []string{"*"}
	} else {
		origins = strings.Split(originsStr, ",")
		for i, origin := range origins {
			origins[i] = strings.TrimSpace(origin)
		}
	}

	handler.CORS = &CORSConfig{
		AllowedOrigins: origins,
		Raw:            value,
	}

	return nil
}

// parseTimeout parses @wylla:timeout 30s
func (p *Parser) parseTimeout(handler *Handler, value string) error {
	// Parse duration string (e.g., "30s", "5m", "1h")
	var num int
	var unit string

	if _, err := fmt.Sscanf(value, "%d%s", &num, &unit); err != nil {
		return fmt.Errorf("invalid timeout format: %s (use format like '30s', '5m', '1h')", value)
	}

	var multiplier int64
	switch unit {
	case "s", "sec", "second":
		multiplier = 1
	case "m", "min", "minute":
		multiplier = 60
	case "h", "hr", "hour":
		multiplier = 3600
	default:
		return fmt.Errorf("invalid timeout unit: %s (use s/m/h)", unit)
	}

	handler.Timeout = parseDuration(int64(num) * multiplier)
	return nil
}

// parseContainerService parses @wylla:container service=name
func (p *Parser) parseContainerService(handler *Handler, value string) error {
	if !strings.HasPrefix(value, "service=") {
		return fmt.Errorf("container parameter must be in format 'service=name', got: %s", value)
	}

	serviceName := strings.TrimPrefix(value, "service=")
	serviceName = strings.TrimSpace(serviceName)

	if serviceName == "" {
		return fmt.Errorf("service name cannot be empty")
	}

	handler.ServiceName = serviceName
	return nil
}

// Helper function to create time.Duration from seconds
func parseDuration(seconds int64) time.Duration {
	return time.Duration(seconds * 1000000000) // Convert to nanoseconds
}

File: typescript/src/annotations/parser.ts
import * as fs from 'fs';
import * as path from 'path';
import {
  Handler,
  ParsedAnnotations,
  ParseError,
  DeploymentType,
  AuthType,
  Route,
  CORSConfig,
  RateLimitConfig,
  TimeoutConfig,
  MemoryConfig,
  ConcurrencyConfig
} from './types';

/**
 * Parser for extracting Box annotations from TypeScript/JavaScript files
 */
export class Parser {
  /**
   * Parse all handlers in a directory
   */
  async parseDirectory(dir: string): Promise<ParsedAnnotations> {
    const result: ParsedAnnotations = {
      handlers: [],
      errors: []
    };

    await this.walkDirectory(dir, result);
    return result;
  }

  /**
   * Parse a single file for handlers
   */
  async parseFile(filePath: string): Promise<ParsedAnnotations> {
    const result: ParsedAnnotations = {
      handlers: [],
      errors: []
    };

    try {
      const content = await fs.promises.readFile(filePath, 'utf-8');
      const handlers = this.extractHandlers(content, filePath);
      result.handlers.push(...handlers);
    } catch (error) {
      result.errors.push({
        filePath,
        lineNumber: 0,
        message: `Failed to read file: ${error}`,
        annotation: ''
      });
    }

    return result;
  }

  /**
   * Recursively walk directory and parse files
   */
  private async walkDirectory(dir: string, result: ParsedAnnotations): Promise<void> {
    const entries = await fs.promises.readdir(dir, { withFileTypes: true });

    for (const entry of entries) {
      const fullPath = path.join(dir, entry.name);

      if (entry.isDirectory()) {
        // Skip node_modules and hidden directories
        if (entry.name === 'node_modules' || entry.name.startsWith('.')) {
          continue;
        }
        await this.walkDirectory(fullPath, result);
      } else if (entry.isFile()) {
        // Only parse TypeScript and JavaScript files
        if (entry.name.endsWith('.ts') || entry.name.endsWith('.js')) {
          // Skip test files and type definition files
          if (!entry.name.endsWith('.test.ts') &&
              !entry.name.endsWith('.test.js') &&
              !entry.name.endsWith('.d.ts')) {
            const fileResult = await this.parseFile(fullPath);
            result.handlers.push(...fileResult.handlers);
            result.errors.push(...fileResult.errors);
          }
        }
      }
    }
  }

  /**
   * Extract handlers from file content
   */
  private extractHandlers(content: string, filePath: string): Handler[] {
    const handlers: Handler[] = [];
    const lines = content.split('\n');

    for (let i = 0; i < lines.length; i++) {
      // Look for function declarations or const assignments
      // Handles: function foo(), export function foo(), const foo = ..., export const foo: Type = ...
      const functionMatch = lines[i].match(/(?:export\s+)?(?:async\s+)?function\s+(\w+)|(?:export\s+)?const\s+(\w+)\s*(?::\s*\w+\s*)?=/);

      if (functionMatch) {
        const functionName = functionMatch[1] || functionMatch[2];

        // Look backwards for annotations
        const annotations = this.extractAnnotationsAbove(lines, i);

        if (annotations.deploymentType) {
          const handler = this.buildHandler(
            functionName,
            filePath,
            annotations,
            i + 1
          );
          if (handler) {
            handlers.push(handler);
          }
        }
      }
    }

    return handlers;
  }

  /**
   * Extract annotations from comments above a function
   */
  private extractAnnotationsAbove(lines: string[], functionLineIndex: number): any {
    const annotations: any = {};

    // Look backwards from function line
    for (let i = functionLineIndex - 1; i >= 0; i--) {
      const line = lines[i].trim();

      // Stop at empty lines or non-comment lines
      if (!line || (!line.startsWith('//') && !line.startsWith('*') && !line.startsWith('/*'))) {
        break;
      }

      // Extract annotation
      const annotationMatch = line.match(/@box:(\w+)\s*(.*)/);
      if (annotationMatch) {
        const [, key, value] = annotationMatch;
        this.parseAnnotation(key, value.trim(), annotations);
      }
    }

    return annotations;
  }

  /**
   * Parse individual annotation
   */
  private parseAnnotation(key: string, value: string, annotations: any): void {
    switch (key) {
      case 'function':
        annotations.deploymentType = DeploymentType.Function;
        break;

      case 'container':
        annotations.deploymentType = DeploymentType.Container;
        break;

      case 'path':
        const pathMatch = value.match(/(\w+)\s+(.+)/);
        if (pathMatch) {
          annotations.route = {
            method: pathMatch[1].toUpperCase(),
            path: pathMatch[2]
          };
        }
        break;

      case 'auth':
        if (value === 'none') annotations.auth = AuthType.None;
        else if (value === 'optional') annotations.auth = AuthType.Optional;
        else if (value === 'required') annotations.auth = AuthType.Required;
        break;

      case 'cors':
        const originsMatch = value.match(/origins=(.+)/);
        if (originsMatch) {
          const origins = originsMatch[1].split(',').map(o => o.trim());
          annotations.cors = { origins };
        }
        break;

      case 'ratelimit':
        const rateLimitMatch = value.match(/(\d+)\/(second|minute|hour|day)/);
        if (rateLimitMatch) {
          const requests = parseInt(rateLimitMatch[1]);
          const period = rateLimitMatch[2] as 'second' | 'minute' | 'hour' | 'day';
          const windowMs = this.periodToMs(period);
          annotations.rateLimit = { requests, period, windowMs };
        }
        break;

      case 'timeout':
        const timeoutMatch = value.match(/(\d+)(s|m|h)/);
        if (timeoutMatch) {
          const duration = parseInt(timeoutMatch[1]);
          const unit = timeoutMatch[2];
          annotations.timeout = {
            duration: this.durationToMs(duration, unit)
          };
        }
        break;

      case 'memory':
        const memoryMatch = value.match(/(\d+)MB/);
        if (memoryMatch) {
          annotations.memory = {
            size: parseInt(memoryMatch[1])
          };
        }
        break;

      case 'concurrency':
        const concurrencyValue = parseInt(value);
        if (!isNaN(concurrencyValue)) {
          annotations.concurrency = {
            max: concurrencyValue
          };
        }
        break;
    }
  }

  /**
   * Build handler object from annotations
   */
  private buildHandler(
    functionName: string,
    filePath: string,
    annotations: any,
    lineNumber: number
  ): Handler | null {
    if (!annotations.route) {
      return null; // Must have a route
    }

    const packageName = path.basename(path.dirname(filePath));

    return {
      packageName,
      functionName,
      filePath,
      deploymentType: annotations.deploymentType,
      route: annotations.route,
      auth: annotations.auth || AuthType.None,
      cors: annotations.cors,
      rateLimit: annotations.rateLimit,
      timeout: annotations.timeout,
      memory: annotations.memory,
      concurrency: annotations.concurrency
    };
  }

  /**
   * Convert period to milliseconds
   */
  private periodToMs(period: 'second' | 'minute' | 'hour' | 'day'): number {
    switch (period) {
      case 'second': return 1000;
      case 'minute': return 60 * 1000;
      case 'hour': return 60 * 60 * 1000;
      case 'day': return 24 * 60 * 60 * 1000;
    }
  }

  /**
   * Convert duration to milliseconds
   */
  private durationToMs(value: number, unit: string): number {
    switch (unit) {
      case 's': return value * 1000;
      case 'm': return value * 60 * 1000;
      case 'h': return value * 60 * 60 * 1000;
      default: return value * 1000;
    }
  }
}
